docs/proposals/20250513-propogate-taints.md

```diff
@@ -0,0 +1,455 @@
+  - Example: Ensure that no workload gets scheduled to a `Node` unless the taint got removed to e.g. install a GPU driver before allowing workload.
+  - CAPI should once set the taint on the Node and not add it again if it got removed.
+
+#### Proposed API changes
+
+##### Type definition of a taint in Cluster API objects
+
+Note: To reduce verbosity, this proposal does not include all kinds of validation markers.
+
+The following defines a new struct which should be used to define taints at the corresponding API types.
+It replicates the upstream corev1.Taint specification and extends it by a field called `propagation`, which will define the propagation mechanism to use for the taint.
+Using a type definition allows to be extensible and add additional propagation mechanisms when necessary.
+
+```golang
+// MachineTaint defines a taint equivalent to corev1.Taint, but additionally having a propagation field.
+type MachineTaint struct {
+  // key is the taint key to be applied to a Node.
+  // +required
+  Key string `json:"key"`
+
+  // value is the taint value corresponding to the taint key.
+  // +optional
+  Value string `json:"value,omitempty"`
+
+  // effect is the effect for the taint. Valid values are NoSchedule, PreferNoSchedule and NoExecute.
+  // +required
+  // +kubebuilder:validation:Enum=NoSchedule;PreferNoSchedule;NoExecute
+  Effect corev1.TaintEffect `json:"effect"`
+
+  // propagation defines how this taint should be propagated to Nodes.
+  // Always: The taint will be continuously reconciled. If removed from a node, it will be re-added.
+  // Initialize: The taint will be added during node initialization. If removed, it will not be re-added.
+  // +required
+  Propagation MachineTaintPropagation `json:"propagation"`
+}
+
+// MachineTaintPropagation defines when a taint should be propagated to Nodes.
+// +kubebuilder:validation:Enum=Always;Initialize
+type MachineTaintPropagation string
+
+const (
+  // TaintPropagationAlways means the taint should be continuously reconciled and kept on the Node.
+  // If removed, it will be re-added during reconciliation.
+  TaintPropagationAlways MachineTaintPropagation = "Always"
+
+  // TaintPropagationInitialize means the taint should be set once during initialization and then
+  // left alone. If removed, it will not be re-added.
+  TaintPropagationInitialize MachineTaintPropagation = "Initialize"
+)
+```
+
+Proper validations on the new field should ensure that no taint with a key of `node.cluster.x-k8s.io/uninitialized` is getting added, because that one is restricted to be used by CAPI and providers.
+
+Open question: alternative to consider, e.g. specific struct, (negative: not as extensible):
+* example
+  ```golang
+  type MachineTaints struct{
+    Always        []corev1.Taint `json:"always,omitempty"`
+    Initialization []corev1.Taint `json:"initialization,omitempty"`
+  }
+  ```
+* FIXME(chrischdi): if decided this should move down to alternative approaches.
```

Comment:
Fine for me to move down. I think one array will look better than two in the YAML